src/SDK/SmartDetectors.Extensions/Presentation/AlertExtensions.cs

```diff
@@ -231,31 +243,101 @@ private static AlertProperty CreateAlertProperty(Alert alert, AlertPresentationP
                     }
 
-                case AlertPresentationSingleColumnTableAttribute singleColumnTableAttribute:
-                    if (!(propertyValue is IList singleColumnTablePropertyValue))
-                    {
-                        throw new ArgumentException("An AlertPresentationSingleColumnTableAttribute can only be applied to properties of type IList");
-                    }
+                case AlertPresentationTableAttribute tableAttribute:
+                    return CreateTableAlertProperty(propertyValue, propertyName, displayName, tableAttribute);
 
-                    return new TableAlertProperty(propertyName, displayName, presentationAttribute.Order, singleColumnTableAttribute.ShowHeaders, singleColumnTablePropertyValue);
+                default:
+                    throw new InvalidEnumArgumentException($"Unable to handle presentation attribute of type {presentationAttribute.GetType().Name}");
+            }
+        }
 
-                case AlertPresentationTableAttribute tableAttribute:
-                    if (!(propertyValue is IList tablePropertyValue))
-                    {
-                        throw new ArgumentException("An AlertPresentationTableAttribute can only be applied to properties of type IList");
-                    }
+        /// <summary>
+        /// Create a new instance of a <see cref="TableAlertProperty{T}"/> based on the given values.
+        /// </summary>
+        /// <param name="propertyValue">The property value, this must be an instance of <see cref="IList{T}"/>.</param>
+        /// <param name="propertyName">The table property name.</param>
+        /// <param name="displayName">The table property display name.</param>
+        /// <param name="tableAttribute">The attribute applied to the table property.</param>
+        /// <returns>The newly created <see cref="TableAlertProperty{T}"/> instance.</returns>
+        private static DisplayableAlertProperty CreateTableAlertProperty(
+            object propertyValue,
+            string propertyName,
+            string displayName,
+            AlertPresentationTableAttribute tableAttribute)
+        {
+            // Validate we have a proper value
+            if (!(propertyValue is IList tablePropertyValue))
+            {
+                throw new ArgumentException("An AlertPresentationTableAttribute can only be applied to properties of type IList");
+            }
+
+            Type tableRowType = GetGenericListType(propertyValue.GetType());
+            if (tableRowType == null)
+            {
+                throw new ArgumentException("An AlertPresentationTableAttribute can only be applied to properties of type IList<>");
+            }
+
+            // Easy way out if we're handling a single-column table
+            if (tableAttribute is AlertPresentationSingleColumnTableAttribute)
```

Comment:
>AlertPresentationSingleColumnTableAttribute [](start = 34, length = 43)

Why do we need this type at all? Can't we just use the multi-column option, even for single-column tables?

Reply:
no we can't, because in the single-column there is no name for the single column, and no property to get it from

---
In reply to: [232756985](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/22#discussion_r232756985) [](ancestors = 232756985)